src/skmatter/neighbors/_sparsekde.py

```diff
@@ -0,0 +1,671 @@
+        kernel: str = "gaussian",
+        verbose: bool = False,
+    ):
+        self.metric_params = (
+            metric_params if metric_params is not None else {"cell_length": None}
+        )
+        self.metric = lambda X, Y: metric(X, Y, squared=True, **self.metric_params)
+        self.cell = metric_params["cell_length"] if metric_params is not None else None
+        self._check_dimension(descriptors)
+        self.descriptors = descriptors
+        self.weights = weights if weights is not None else np.ones(len(descriptors))
+        self.weights /= np.sum(self.weights)
+        self.fspread = fspread
+        self.fpoints = fpoints
+        self.kernel = kernel
+        if self.kernel != "gaussian":
+            raise NotImplementedError
+        self.verbose = verbose
+
+        if self.fspread > 0:
+            self.fpoints = -1.0
+
+        self.bandwidth_ = None
+        self._covariance = None
+
+    @property
+    def nsamples(self):
+        return len(self.descriptors)
+
+    @property
+    def ndimension(self):
+        return self.descriptors.shape[1]
+
+    @property
+    def kdecut_squared(self):
+        return (3 * (np.sqrt(self.descriptors.shape[1]) + 1)) ** 2
+
+    @property
+    def _bandwidth_inv(self):
+        if self.fitted_:
+            if self._bandwidth_inv_ is None:
+                self._bandwidth_inv_ = np.array(
+                    [np.linalg.inv(h) for h in self.bandwidth_]
+                )
+        else:
+            raise ValueError("The model is not fitted yet.")
+        return self._bandwidth_inv_
+
+    @property
+    def _normkernels(self):
+        if self.fitted_:
+            if self._normkernels_ is None:
+                self._normkernels_ = np.array(
+                    [
+                        self.ndimension * np.log(2 * np.pi) + np.linalg.slogdet(h)[1]
+                        for h in self.bandwidth_
+                    ]
+                )
+        else:
+            raise ValueError("The model is not fitted yet.")
+        return self._normkernels_
+
+    def fit(self, X, y=None, sample_weight=None):
+        """Fit the Kernel Density model on the data.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
+            corresponds to a single data point.
+
+        y : None
+            Ignored. This parameter exists only for compatibility with
+            :class:`~sklearn.pipeline.Pipeline`.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            List of sample weights attached to the data X. This parameter
+            is ignored. Instead of reading sample_weight from the input,
+            it is calculated internally.
+
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        self._bandwidth_inv_ = None
+        self._normkernels_ = None
```

Comment:
I think it would be more transparent to the reader of the code if you set `self._bandwidth_inv = None`. Otherwise in the remaining code you don't use it anymore and it becomes only clear when reading the property. 

 For that you need a property setter
```python
@_bandwith_inv.setter
def _bandwith_inv(self, value)
    self._bandwith_inv_ = value
```
Alternatively, you could use https://docs.python.org/3/library/functools.html#functools.cached_property then you can reset cached_properties with `del self.property_name`.

Same for `_normkernels_`

Reply:
After consideration, I think adding some comments at these lines might be better. There are two reasons why I did not take the two ways you mentioned. For the setter way, it does allow me to re-set the value of `_bandwith_inv` much clearer, but it also exposes the logic of calculating it. For the `cached_property` way, the value of these two properties won't be updated if the user do the fitting twice.

But I am not pretty sure if not exposing the logic of setting these two properties makes sense. If not, I think the setter way is better.

Reply:
> For the cached_property way, the value of these two properties won't be updated if the user do the fitting twice.

If you set it to None at the beginning, as you do it anyway here, they can be updated. I think that would be the better solution here, but I am fine with the comment.